### ISSUE TITLE
Enable Bulk Printing of Invoices destined for specific Shipping Method only

### ISSUE THREAD

**Issue by lbwright22** [openfoodfoundation/openfoodnetwork#5277]
## Description

As a manager of a busy hub which has multiple shipping methods (ie different collection points, delivery routes and times) I would like to filter my orders from that weeks order cycle and print the invoices for each shipping method in bulk separately. 
By being able to do so, I can then keep the paper work separate. This will help with putting the correct paper work with a customer's order when packing for each different drop off or delivery point.  It would save a lot of time.

To do this I would like on the Admin-> Orders page, where all the orders are listed, to be able to sort orders by Shipping Method as well as Distributor and Order Cycle.
Once only those orders associated with a given shipping method are present, I can 'Select All' and print invoices in bulk.

## Acceptance Criteria & Tests

1. At the end of an order cycle, visit Admin -> Orders page
2. Filter Orders from Distributor A, Order Cycle AA and with Shipping Method AAA
3. Select All which fit this criteria
4. Select Print Invoices
5. Repeat for Shipping Methods BBB, CCC etc

**Comment by lbwright22** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Invoices & Pick Lists](https://trello.com/c/Vn1V1M9M/912-invoices-pick-lists)

**Comment by luisramos0** [comment]
We currently have filters for OC and distributor and the table only contains distributor, not OC.
You can add sorting to the distributor column, papercut.

Do you want to add two more columns to the table?  OC and ship method? Will that not make the table too complex/wide?
Has that been agreed with the product team?

UX wise, for simplicity, maybe the best solution would be to add just a filter for ship method and keep the columns in the table of results as is? 

Can you add the tag enhacement to the papercut candidate you create? thanks!

**Comment by lin-d-hop** [comment]
I would suggest that this is simply changing the ability to filter - not the columns in the table.

Just like you can filter by OC - we want to be able to filter by shipping method.

@lbwright22 I think you can simplify this story somewhat. Here we don't really need to know about Invoice printing at all. We simply want to be able to filter on the Orders page by Shipping Method. Perhaps the story would be clearer if we specifically state:
- That we don't want the display columns to change in this story.
- That we want the Shipping Method to filter

You could add a note: This addition will enable hub managers to print invoice batches per Shipping Method.

**Comment by jeduardo824** [comment]
Hey guys, I checked that. It looks like it will be easy to accomplish, but I'm having trouble getting the shipping methods list. I tried with `Spree::ShippingMethod.managed_by(current_spree_user)` but it returns some duplicated entries. Anyone knows a better way to get that?

**Comment by luisramos0** [comment]
that sounds correct @jeduardo824 
It's the method used in Subscriptions.
can you show the list of ship methods? are you sure there are duplicates?
anyway, I think you can go and fix the scope to return a unique list of ship methods.

**Comment by jeduardo824** [comment]
![Screen Shot 2020-05-03 at 09 12 36](https://user-images.githubusercontent.com/27960597/80913873-52729e00-8d1e-11ea-8e86-6ad612bb3c35.png)

**Comment by luisramos0** [comment]
that means the names are repeated, not that ship methods are repeated right? can you please verify the ids?

**Comment by jeduardo824** [comment]
![Screen Shot 2020-05-03 at 09 28 24](https://user-images.githubusercontent.com/27960597/80914207-78993d80-8d20-11ea-9b58-0dc4a6625842.png)

**Comment by jeduardo824** [comment]
They have different ids

**Comment by jeduardo824** [comment]
So it looks like the shipping methods are different but with the same name, correct?

**Comment by luisramos0** [comment]
yes, ok.
you are super admin and probably using sample data?
The same 2 ship methods are created for each enterprise lib/tasks/sample_data/shipping_method_factory.rb

So, all good :+1:

**Comment by luisramos0** [comment]
I fixed this problem in this new PR: https://github.com/openfoodfoundation/openfoodnetwork/pull/5351

**Comment by jeduardo824** [comment]
Awesome! Also, just a last question because this is the first I'm dealing with Ransack. Do you think I need to make DB changes in Spree::Order like add the shipping_method_id to the able or I'm able to accomplish that using Arel?

**Comment by luisramos0** [comment]
no, thank god you dont need to change the data model! But this is not an easy task...

you need to make sure the shipping methods table is included in the query and then your filter is also in the query.

You need to understand ransack, it's pretty easy, have a look here: https://github.com/activerecord-hackery/ransack
it's pretty easy!

I think you need to add a q param in
app/assets/javascripts/admin/orders/controllers/orders_controller.js.coffee line 39:
      'q[shipping_method_id_eq]': $scope['q']['shipping_method_id_eq'],
and make it reach the server with the value selected by the user.

Then on the server, through Api::OrdersController#index it will get automatically to SearchOrders that uses private fetch_orders where params[:q] is already used.
All you need to do is make sure the Permissions::Order.editable_orders includes the shipping method... **this is the tricky part**.

This is not trivial because to get the shipping method of an order in the DB you have to go through order.shipments.shipping_rates.selected.shipping_method_id :see_no_evil: 

Basically something like this:
/home/luisramos/dev/docker/ofn/lib/open_food_network/order_cycle_management_report.rb#111
orders.joins(shipments: :shipping_rates).where(spree_shipping_rates: { shipping_method_id: params[:shipping_method_in] })

But I think this has a bug... because a shipping_method can have a shipping_rate in an order that is not selected. The order shipping_method is given by spree_shipping_rates where field selected is true.

I'd try to add the tables in SearchOrders.fetch_orders
@search = ::Permissions::Order.new(current_user).editable_orders.
**joins(shipments: :shipping_rates).where(spree_shipping_rates: { selected: true }})**
.ransack(params[:q])

will ransack do the rest? It will have to match the ship method id in params[:q] to spree_shipping_rates.shipping_method_id

I am sorry @jeduardo824, this may not be easy for someone who is not very familiar with OFN!

**Comment by jeduardo824** [comment]
Well, I already accomplished half of what you said. I was trying now to join the shipping_method.  Your clue about the shipping_rate is precious. I will try to finish it with the tips you gave. Thanks!

**Comment by luisramos0** [comment]
awesome, please submit a PR even if you cant make it work correctly.
